### PR TITLE
Feature/updated design

### PR DESCRIPTION
**Pull request by HonoreDB**
@sheavalentine-mdsol @csavage-mdsol This is my proposed minimum viable API for Farscape R3. Have I taken out, or failed to add, anything we need? Note that we're still exposing the representor hash, as before...the API described here is just about entry points and hardcoded workflows.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -134,23 +122,42 @@ new_drd.transitions # => ['self', 'edit', 'delete', 'deactivate', 'leviathan']
...
 ```
 
-With bookmarks, you can decorate links in UI applications to make it simple to enter state on the backend when 
-interacting with the UI links and forms.
+Request parameters can be passed as a hash or as a block.
+
+```ruby
+# The following are all equivalent:
+
+drd = drds.create!(name: 'Pike')
```
Comment (csavage-mdsol) [inline README.md:155]: Do you see developers using `respond_to?` a fair amount with the railsy interface?  Actually, that could get slightly hairy.

Comment (HonoreDB) [inline README.md:155]: I suppose they might end up needing to, yeah. Is there an extra concern there vs. say `transitions['create'].try(:invoke)` or `transitions.key? 'create'` ?

Comment (csavage-mdsol) [inline README.md:155]: I think it increases the chances of collision.  If plinth had named `lock` for studies `freeze` for example.  Sort of reminds me of the dangers of `dup`, like this being busted:

``` ruby
value.respond_to?(:dup) ? value.dup : value
```

Comment (csavage-mdsol) [inline README.md:155]: Oh, config type clone is a better example

Comment (HonoreDB) [inline README.md:155]: Definitely an issue. I mention it in line 159.

Comment (csavage-mdsol) [inline README.md:155]: ah, I did not get that far before I decided to stick my foot squarely in my mouth.

**Comment by fosrias** [comment]
@HonoreDB @sheavalentine-mdsol @csavage-mdsol @isaacwong One comment I would make as you think about this interface is that you do not get the benefit of a hypermedia API if you use the client in an RPC way. The loose coupling and evolvability will come from being paranoid. You are NOT interacting with a resource or remote anything conceptually. You are introspecting and interacting with a client side, state-machine message. So, to get the loose coupling and evolvability you have to assume nothing and always check before taking action.

The moment you go the metaprogramming, railsy way you and just assume stuff will be there, you lose the benefit. You may want to think about helper predicate methods so that you can do this (you should be paranoid about assuming properties are there too, not just transitions).

``` ruby
if message.create?
  message.create ...
else
  # I code in my recovery path.
end
```

This is how you have to write against hypermedia APIs. There are other way, maybe returning NullRepresentor or other patterns, but that ideas in that simple if block is the vast difference between traditional HTTP clients and what you are trying to do. FWIW.

**Comment by brutski** [comment]
@fosrias Isn't it exact same content you sent in email the day you left? If I remember right, we agreed that 

``` ruby
transition = api.get_transition('get')
if transtion
#do something
else
#recover
end
```

is way better than just:

``` ruby
transition = api.invoke('get')
```

It was email with something about horror and you not sleeping at night because of RPC we invented in farscape.

**Comment by fosrias** [comment]
@brutski Rings a bell.

**Comment by HonoreDB** [comment]
Thanks for weighing in. I definitely agree that "hardcoded" consumption like this should be the exception, not the rule, and it should be written in the awareness that it's more likely to break than a similar integration with an API with a published contract.

In cases where it's necessary, I'm not thrilled with the idea of having similar conditionals wrapping every call. Maybe we should DRY it with a NullRepresentor, like you say, or with something like

``` ruby
agent.on_missing { default recovery code }

drds.deactivate! => # calls the block for the agent
drds.deactivate! { custom recovery code } => #Overrides the default recovery behavior
```

**Comment by fosdev** [comment]
Yeah. I was talking with Mike Amundsen, and he was telling me how he was pondering a hypermedia friendly language because he was writing so many if … else blocks. The challenge is that you are going to have to go out of your way to break how developers currently think about consuming APIs. Using some kind of recovery blocks like you suggest or some creativity is in order. Knee jerk will be to think euresourcey, ActiveResourcey, which really brittle anti patterns of API clients because they effectively act like they are tunneling remote objects.

In any case, will be interested to see how this turns out.

**Comment by sheavalentine-mdsol** [comment]
Some thoughts...

# Initialization

``` ruby
agent = Farscape::Agent.new('http://example.com/my_api')
resources = agent.enter
```

is there a reason not to also allow

``` ruby
agent = Farscape::Agent.new

resources = agent.enter('http://example.com/my_api')
```

# Regarding 'enter'

Why/how is enter different from invoke?  Isn't an entry point just a resource that communicates resources?  Is there a reason for the API to be different?

# Applying Query Parameters

What about uri templating? 

Why is the developer choosing between attributes and parameters?  Presumably the mediatype specifies where each of those go.  In that case why not just dump all the information you know into the invocation and have Farscape sort it out?

Also why can't a developer do

``` ruby
filtered_drds = search_transition.invoke { search_term: '1812' }
#or even
filtered_drds = search_transition.call { search_term: '1812' }
#or
filtered_drds = search_transition.( { search_term: '1812' } )
```

Another thought, why not:

``` ruby
filtered_drds = drds.transitions.('search', { search_term: '1812' })
#or
filtered_drds = drds.transitions.call('search', { search_term: '1812' })
```

Finally, it seems that the ideal case would involve something like:

``` ruby
search_transition.parameters # => ['search']

# utilizing the profile
search_transition.parameters.semanticly_bound # => ['search_term']

known_data = {
   search_term: '1812'
   force_filed: true
}

# equivalent to filtered_drds = search_transition.invoke { search: '1812' }
filtered_drds = search_transition.invoke known_data
```

# Alternate interface

How does Farscape differentiate between:

``` ruby
document = {
  leviathon: '1234',
  _links: {leviathon: {href: '...'}}
}
drd.leviathan
```

It seems to me if you really want to go down this route (which I'm unconvinced is a good idea) you should do this:

``` ruby
drd['leviathan'] # => 1234
drd[:leviathan] # => 1234
drd.leviathan # => Equivalent to drd.transitions['leviathan'].invoke
```

Though even here I'd actually recommend:

``` ruby
drd['leviathan'] # => 1234
drd[:leviathan] # => 1234
drd.leviathan # => Equivalent to drd.transitions['leviathan']
drd.leviathan.call # => Equivalent to drd.transitions['leviathan'].invoke
drd.leviathan.() # => Equivalent to drd.transitions['leviathan'].invoke
```

**Comment by sheavalentine-mdsol** [comment]
How is it going to deal with a list of transitions with a single link relation?
Similarly how does it differentiate between that list of transition and their equivalent embedded item in the 'alternate' interface?

**Comment by HonoreDB** [comment]
_Initialization and Entry_
My idea here is to discourage using Farscape as a fancy curl, passing in URLs instead of transitions. Setting an entry point only on initialization emphasizes that you only input a URL once.

_Applying Query Parameters_
I'll have to look more into how attributes relate to parameters here. The API you suggest seems functionally equivalent to my alternate interface--handling semantic binding would be the job of the builder object in the block form.

_Alternate interface_
I hadn't thought of that collision. Having to disambiguate it makes me sad, because I liked the thought of the client not needing to know whether something is an attribute or a transition--it lets a service change an attribute into a fully qualified child resource whenever it wants. Maybe your `[:leviathan]` syntax can always refer to an attribute, and the `.leviathan` form refers to a transition if it exists, otherwise an attribute. So when we know there's a collision, we can access both, otherwise be can be agnostic.

**Comment by sheavalentine-mdsol** [comment]
Initialization and Entry
Can it be started with any URL, or only some sort of "entry point"?

Alternate interface
I don't like the idea that it's sometimes one thing and sometimes another.  A method and a transition both involve action so the metaphor seems to fit there.  Data access seems to map more closely to a hash interface, so that seems sensible there as well.

"I liked the thought of the client not needing to know whether something is an attribute or a transition"
I'm trying to imagine how this could be a good thing.  When you access an attribute you are going to have a different API than when you access a transition.  That means they don't ducktype at all, so after you access it you'll need to worry about making sure it's one sort of thing.

The Alternate Interface is railsy?  Can you point to what you're inspiration is?

**Comment by sheavalentine-mdsol** [comment]
What happens when you deserialize this with farscape?

``` json
{
"_links": {
  "allocate": {...}, 
  "new": {...}, 
  "superclass": {...}, 
  "freeze": {...}, 
  "to_s": {...}, 
  "constants": {...}, 
  "class_variables": {...},
  "instance_method": {...}, 
  "hash": {...}, 
  "inspect": {...}, 
  "methods": {...}, 
  "method": {...}, 
  "object_id": {...}
}
"total_count": 0
```

**Comment by HonoreDB** [comment]
We don't have a way of distinguishing between entry points and non-entry points right now, as far as I know. The README as is talks about using different media types for the two but that seems extreme to me.

I suspect attributes and invoked-transitions are going to be somewhat duck-typable if we just continue the pattern of giving them similar interfaces and keep it in mind as a design goal.

Railsiness of the Alternate Interface: This is how ActiveRecord::Base works. It queries the database schema, and converts columns to attributes and foreign keys to methods that call the database. You can pass `@user.study` to a view without knowing whether `study` is a string or a Study model instance. The pattern is generalized in [ActiveResource](https://github.com/rails/activeresource).

I cover conflicts in the README lower down, the preexisting method wins so you have to use the other interface. Which is again how Rails often handles that.

**Comment by sheavalentine-mdsol** [comment]
Proposal:

``` ruby
resource.transitions # => {:hash, 'of transitions'}
resource.embedded # => {hash: 'of embedded'}
resource.attributes # => {:hash, 'of attributes'}
resource.headers # => {:hash, 'of request headers'}
resource.meta # => {:hash, 'of metadata'}

#attribute access
resource.attributes[:key] # => returns a field object (see representor)
resource.attributes[:key].value # => returns a field value
resource.attributes[:key].call # => returns a field value
resource[:key] # => returns a field value

#transition access
resource.transitions[:key] # => returns a transition object
resource.transitions[:key].invoke do |builder|
  builder.parameters = { search_term: '1812' }
end # => dereference link and return resource
resource.transitions[:key].call { search_term: '1812' } # => dereference link and return resource
resource._key # => alias for resource.transitions[:key]

#if you wanted to get the '_doit' transition, you'd use
resource.__doit # => alias for resource.transitions[:_doit]

#No namespace conflict possible
```

**Comment by sheavalentine-mdsol** [comment]
``` ruby
embedded_drd_items = drds.items
```

This doesn't make sense.  with 'new interface' _.items would be an items attribute.
This should be 

``` ruby
embedded_drd_items = drds.embedded['items']
```

or something...

**Comment by sheavalentine-mdsol** [comment]
Also @csavage-mdsol  and I were talking about the problem of 'fixed' links and having to check for it's existence first.  I proposed we could wrap the return from method_missing in a Maybe monad, and Connor suggested that is pretty much the railsy way with Try

**Comment by HonoreDB** [comment]
`items` is just another namespace collision, we'll handle it the same way.

I like the Maybe pattern in general, I do think it could work well here.

**Comment by sheavalentine-mdsol** [comment]
.items is fine with the alternate interface but the part where it's talking about it is with the 'standard' interface.

**Comment by sheavalentine-mdsol** [comment]
agent.enter({rel}) in the spec is a URL because it's using the weird entry point rels.  It should be a standard rel for clarity.

File: README.md [hunk 2]
```diff
@@ -50,15 +43,10 @@ drds.attributes # => { total_count: 25, items: [...] }
 drds.transitions.keys # => ['self', 'search', 'create', 'next', 'last']
 
 agent.enter('http://example.com/rel/unknown_resource') # => raises Farscape::Agent::UnknownEntryPoint
```
Comment (sheavalentine-mdsol) [inline README.md:45]: I do not understand the need for a separate entry interface, and it seems to complicate things.
This makes sense

``` ruby
Farscape::Agent.new(entry_point).enter
Farscape::Agent.new.enter(entry_point)
Farscape::Agent.new.enter # Raise Error
```

However this seems like WAT!

``` ruby
agent = Farscape::Agent.new(entry_point).enter.transitions['drds'].invoke
agent = Farscape::Agent.new(entry_point).enter("#{entry_point}/drds")
agent = Farscape::Agent.new(entry_point).enter("unknown_resource") # Raises Error
```

**Comment by miwest929** [comment]
@HonoreDB What's the deal with this PR? Should it be closed in favor of another one?

**Comment by HonoreDB** [comment]
We're just using it as a discussion thread, really, but it can get merged in once the interface is implemented so its content becomes true.

**Comment by miwest929** [comment]
@HonoreDB Sounds good.

**Comment by sheavalentine-mdsol** [comment]
Open a new PR against Brave New Develop?

**Comment by HonoreDB** [comment]
Makes sense.

**Pull request closed without merging**